src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -166,139 +244,109 @@ public static Response Apply(Response current, Stack<Request> requests, List<Res
+
+        public ReplaceTypeParamCallsExpressionKind(ReplaceTypeParamCallsExpression expr,
+            Dictionary<QsTypeParameter, ResolvedType> currentParamTypes,
+            GetConcreteIdentifierFunc getConcreteIdentifier) : base(expr)
         {
-            Concretion result = new Concretion();
-            foreach (var kvp in typeDict)
-            {
-                result.Add((kvp.Key, kvp.Value));
-            }
-            return result;
+            GetConcreteIdentifier = getConcreteIdentifier;
+            CurrentParamTypes = currentParamTypes;
         }
 
-        private Identifier GetConcreteIdentifier(Identifier.GlobalCallable globalCallable, ImmutableDictionary<QsTypeParameter, ResolvedType> types)
+        public override QsExpressionKind<TypedExpression, Identifier, ResolvedType> onIdentifier(Identifier sym, QsNullable<ImmutableArray<ResolvedType>> tArgs)
         {
-            Concretion target = null;
-            if (types != null && !types.IsEmpty)
+            if (sym is Identifier.GlobalCallable global)
             {
-                target = CreateConcretion(types);
-            }
+                Concretion applicableParams = CurrentParamTypes
+                    .Where(kvp =>
+                        kvp.Key.Origin.Namespace.Value == global.Item.Namespace.Value &&
+                        kvp.Key.Origin.Name.Value == global.Item.Name.Value)
+                    .ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value);
 
-            string name = null;
-
-            // Check for recursive call
-            if (CurrentResponse.originalName.IsEqual(globalCallable.Item) &&
-                CurrentResponse.typeResolutions.SetEquals(target))
-            {
-                name = CurrentResponse.concreteCallable.FullName.Name.Value;
+                if (applicableParams.Any())
+                {
+                    // Create a new identifier
+                    sym = GetConcreteIdentifier(global, applicableParams);
+                    tArgs = QsNullable<ImmutableArray<ResolvedType>>.Null;
+
+                    // Remove Type Params used from the CurrentParamTypes
+                    foreach (var key in applicableParams.Keys)
+                    {
+                        CurrentParamTypes.Remove(key);
+                    }
+                }
             }
+            return base.onIdentifier(sym, tArgs);
+        }
+    }
 
-            // Search Requests for identifier
-            if (name == null)
-            {
-                name = Requests
-                    .Where(req =>
-                        req.originalName.IsEqual(globalCallable.Item) &&
-                        req.typeResolutions.SetEquals(target))
-                    .Select(req => req.concreteName.Name.Value)
-                    .FirstOrDefault();
-            }
+    internal class ReplaceTypeParamCallsExpressionType : ExpressionTypeTransformation<ReplaceTypeParamCallsExpression>
```

Comment:
It seems like the a lot cleaner solution for this transformation is to just not use MinorTransformations.ReplaceTypeParams (please make that internal again) and override onTypeParameter here.